### PR TITLE
Ssh rust v8

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3445

Describe changes:
- Translates SSH C Parser to a Rust parser

This looks good

This is meant to be an exact translation. We want to add features later.
https://github.com/rusticata/ssh-parser might help for this

One difference though is the handling of end of line(s) at the end of the banner. This Rust version should be robust against packet splitting between CR and LF (which should be accepted by openssh), when the current version is vulnerable because end of lines skipping https://github.com/OISF/suricata/blob/master/src/app-layer-ssh.c#L235 is only done if the first packet contains the whole banner with end of lines. I made a suricata-verify test for this https://github.com/OISF/suricata-verify/pull/176

Modifies #4667 with alert-prelude.c fix

### CODE REVIEW

File: rust/src/ssh/ssh.rs [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+    }
+
+    fn set_event(&mut self, event: SSHEvent) {
+        let ev = event as u8;
+        core::sc_app_layer_decoder_events_set_event_raw(&mut self.transaction.events, ev);
+    }
+
+    fn parse_record(&mut self, mut input: &[u8], resp: bool) -> bool {
+        //TODO add SCLogDebug ?
+        let mut hdr = &mut self.transaction.cli_hdr;
```
Comment (victorjulien) [inline rust/src/ssh/ssh.rs]: Could be written as:
```rust
let mut hdr = if !resp {
    &mut self.transaction.cli_hdr
} else {
    &mut self.transaction.srv_hdr
}
```

File: rust/src/ssh/ssh.rs [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+        if resp {
+            hdr = &mut self.transaction.srv_hdr;
+        }
+        //first skip record left bytes
+        if hdr.record_left > 0 {
```
Comment (victorjulien) [inline rust/src/ssh/ssh.rs]: I'm finding this buffering logic quite hard to follow. Could we somehow simplify it?

File: rust/src/ssh/ssh.rs [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+                vec.extend(input);
+                iline = &vec[..];
+            }
+        }
+        match parser::ssh_parse_line(iline) {
+            Ok((rem, line)) => {
+                let mut set_event_long = false;
+                if hdr.banner.len() + line.len() <= SSH_MAX_BANNER_LEN {
+                    hdr.banner.extend(line);
+                } else if hdr.banner.len() < SSH_MAX_BANNER_LEN {
```
Comment (victorjulien) [inline rust/src/ssh/ssh.rs]: In what case would a normal `} else {` not be sufficient here?

File: src/alert-prelude.c [hunk 4]
```diff
@@ -724,25 +724,24 @@ static void PacketToDataProtoTLS(const Packet *p, const PacketAlert *pa, idmef_a
...
     json_t *js, *s_js;
-    SshState *ssh_state = (SshState *)FlowGetAppState(p->flow);
+    void *ssh_state = FlowGetAppState(p->flow);
 
     if (ssh_state == NULL)
         return;
 
-    js = json_object();
-    if (js == NULL)
+    void *tx_ptr = rs_ssh_state_get_tx(ssh_state, 0);
```
Comment (victorjulien) [inline src/alert-prelude.c]: should we test for failure here?

File: src/alert-prelude.c [hunk 5]
```diff
@@ -724,25 +724,24 @@ static void PacketToDataProtoTLS(const Packet *p, const PacketAlert *pa, idmef_a
...
-
     s_js = json_object_get(js, "server");
     if (s_js != NULL) {
         JsonToAdditionalData(NULL, s_js, alert);
     }
 
     s_js = json_object_get(js, "client");
     if (s_js != NULL) {
-        JsonToAdditionalData(NULL, s_js, alert);
+            JsonToAdditionalData(NULL, s_js, alert);
```
Comment (victorjulien) [inline src/alert-prelude.c]: indent?

File: src/util-lua-ssh.c [hunk 6]
```diff
@@ -95,13 +99,17 @@ static int GetServerSoftwareVersion(lua_State *luastate, const Flow *f)
...
 
-    SshState *ssh_state = (SshState *)state;
+    const uint8_t *software = NULL;
+    uint32_t b_len = 0;
 
-    if (ssh_state->srv_hdr.software_version == NULL)
-        return LuaCallbackError(luastate, "error: no server software version");
+    void *tx = rs_ssh_state_get_tx(state, 0);
+    if (rs_ssh_tx_get_software(tx, &software, &b_len, STREAM_TOCLIENT) != 1)
+        return LuaCallbackError(luastate, "error: no server proto version");
```
Comment (victorjulien) [inline src/util-lua-ssh.c]: software version

File: src/util-lua-ssh.c [hunk 7]
```diff
@@ -157,13 +169,17 @@ static int GetClientSoftwareVersion(lua_State *luastate, const Flow *f)
...
 
-    SshState *ssh_state = (SshState *)state;
+    const uint8_t *software = NULL;
+    uint32_t b_len = 0;
 
-    if (ssh_state->cli_hdr.software_version == NULL)
-        return LuaCallbackError(luastate, "error: no client software version");
+    void *tx = rs_ssh_state_get_tx(state, 0);
+    if (rs_ssh_tx_get_software(tx, &software, &b_len, STREAM_TOSERVER) != 1)
+        return LuaCallbackError(luastate, "error: no server proto version");
```
Comment (victorjulien) [inline src/util-lua-ssh.c]: client software

File: rust/src/ssh/ssh.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: I would love to but did not find a Rust way to do it (I arrived to the same code for HTTP2)

Problem comes from the fact that the buffer belongs to the `SSHState` structure
And borrow checker complains if we use the structure in another way when we use/read/reset the buffer

Do you have some pointer to a better code ?

File: rust/src/ssh/ssh.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: ok, thanks for the Rust idiomatics

File: rust/src/ssh/ssh.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: I just want to set the event once, and we do not need to `extend`  if the buffer is already full

File: src/alert-prelude.c (hunk 4 above)
Comment (catenacyber) [reply src/alert-prelude.c]: Well, for SSH, if there is a not NULL state, there is one and only one transaction (see implementation of `rs_ssh_state_get_tx`).

In the C parser, there was only one structure for both state and transaction.

So, should we `BUG_ON` to have no performance cost on production and enforce this condition during tests ?

File: rust/src/ssh/ssh.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/ssh/ssh.rs]: I think the DNS code takes a simpler approach. But, do we still need this after the 'incomplete' logic of #4677 is merged?

File: src/alert-prelude.c (hunk 5 above)
Comment (catenacyber) [reply src/alert-prelude.c]: Ok, thanks

File: src/util-lua-ssh.c (hunk 6 above)
Comment (catenacyber) [reply src/util-lua-ssh.c]: ok thanks

File: src/util-lua-ssh.c (hunk 7 above)
Comment (catenacyber) [reply src/util-lua-ssh.c]: ok thanks

File: src/alert-prelude.c (hunk 4 above)
Comment (victorjulien) [reply src/alert-prelude.c]: `BUG_ON` sounds good to me.

Comment (catenacyber) [reply src/alert-prelude.c]: ok for `BUG_ON` then

File: rust/src/ssh/ssh.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: #4677 will definitely help. So I can rebase when it gets merged

Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: As for DNS, the simpler code has a performance cost of slice allocation + memory copy for each call to the parser cf https://github.com/OISF/suricata/blob/master/rust/src/dns/dns.rs#L523

Can we avoid this tradeoff and have the best of both worlds ?

Comment (victorjulien) [reply rust/src/ssh/ssh.rs]: This cost seems small, but lets just wait for #4677 so we can remove the buffering both here and from DNS.

Comment (catenacyber) [reply rust/src/ssh/ssh.rs]: ok

**Pull request closed without merging**